packages/grid/x-data-grid/src/components/cell/GridCell.tsx

```diff
@@ -249,7 +250,7 @@ const GridCell = React.forwardRef<HTMLDivElement, GridCellProps>((props, ref) =>
     sectionIndex === 0;
   const showRightBorder =
     rootProps.showCellVerticalBorder &&
-    !(isSectionLastCell && pinnedPosition !== PinnedPosition.LEFT);
+    (isSectionLastCell ? pinnedPosition === PinnedPosition.LEFT || !gridHasScrollX : true);
```

Comment:
@romgrk 
This fixes the `ColumnSpanningNumber` demo: https://deploy-preview-10059--material-ui-x.netlify.app/x/react-data-grid/column-spanning/#number-signature